### ISSUE TITLE
Generic link provider

### ISSUE THREAD

**Issue by rodolf0** [xtermjs/xterm.js#2893]
The web link add-on allows matching a restricted set of text (eg. Http links) and lifting that into an object you can interact with.

Would it be possible to create a slightly more general version of it? It would takes 2 construction parameters:
- match: a string defining a Regex Wich allows matching specific text snippets. This Regex should allow capture groups.
- strategy: a string used as a template for rendering a rewrite of the matched snippet optionally using the capture groups from the match definition.

**Comment by Tyriar** [comment]
That's essentially what the old API can accomplish:

https://github.com/xtermjs/xterm.js/blob/939de5d1510123702b5c7c2f36822d19e25f2b66/typings/xterm.d.ts#L730-L750

But this one's on the way out, there's been a bunch of work done recently on the "link provider" API that gets asks for lines at a particular cell:

https://github.com/xtermjs/xterm.js/blob/939de5d1510123702b5c7c2f36822d19e25f2b66/typings/xterm.d.ts#L752-L758

The link provider API will probably soon ask for all links on a particular line, see https://github.com/xtermjs/xterm.js/issues/2848

**Comment by rodolf0** [comment]
Are there other link provider addons available beyond the web links one? Main intention behind this request was to make it easier to craft new link providers with just {match, replacement template} instead of having to build a full addon for each snippet you want to match.

**Comment by Tyriar** [comment]
Only the web one is in this repo, there are more examples in the vscode repo though:

https://github.com/microsoft/vscode/tree/master/src/vs/workbench/contrib/terminal/browser/links

Isn't the link matcher API basically what you're after? I'm not sure exactly what you had in mind with strategy, but that API lets you set a regex, matchIndex to use and then a handler (which is needed for most cases so it's mandatory). Perhaps we should tweak that into a `registerRegexLinkProvider` so you can have the best of both worlds? This would go a long way towards solving https://github.com/xtermjs/xterm.js/issues/2850 as the complexity around wide and emoji character would be hidden for that particular API.

**Comment by rodolf0** [comment]
> Isn't the link matcher API basically what you're after? I'm not sure exactly what you had in mind with strategy, but that API lets you set a regex, matchIndex to use and then a handler (which is needed for most cases so it's mandatory). Perhaps we should tweak that into a `registerRegexLinkProvider` so you can have the best of both worlds? This would go a long way towards solving #2850 as the complexity around wide and emoji character would be hidden for that particular API.

I do see that this is what the original API being deprecated allows.

I think a `RegexLinkProvider` addon would be pretty great! I think the word "strategy" may be confusing maybe we should call these "match" and "template" (instead of strategy).

Let me give some examples of what I'm after.

- Being able to click into gitlab bugs
  - convert 'GNOME/gtk#266' -> 'https://gitlab.gnome.org/GNOME/gtk/issues/266'
  - match '((?i)gnome|community|world)/([-+./a-z]+)#([[:digit:]]+)'
  - template 'https://gitlab.gnome.org/$1/$2/issues/$3'

- Click through Phabricator Diffs (or Tasks, etc.)
  - convert 'D23432432' -> 'https://yourdomain.com/diff/D23432432/'
  - match 'D[0-9]{5,10}'
  - template 'https://yourdomain.com/diff/$0/'


This addon allows matching different captures of the regex. Maybe the core idea is `uri = "some string snippet in the terminal".replace(Regex(match), template); callback(uri)`

**Comment by Tyriar** [comment]
Oh ok that makes sense. This new API would also need to surface these APIs so embedders could use them to give underline feedback, modifier clicks and tooltips in a consistent way:

https://github.com/xtermjs/xterm.js/blob/1168c6c52f49c7155094fe173ac037ea0e9ce8ac/typings/xterm.d.ts#L1124-L1152

**Comment by Tyriar** [comment]
Maybe there should be a new class in the API instead:

```ts
class RegexLinkProvider implements ILinkProvider {
...
}
```

That way there's no need to worry about exposing/duplicating `ILinkProvider` members

**Comment by jerch** [comment]
+1 for the regexp link provider idea.

@Tyriar If we spawn things directly from `LinkProvider` - do we get into "first come/matched - first served" issues during the linkifier provider eval? (Sounds like a classical router problem to me, where registering order gets important...)

**Comment by Tyriar** [comment]
@jerch yes order is important, but the order is defined by the embedder so all good.

**Comment by rodolf0** [comment]
The constructor for `RegexLinkProvier` could take an additional `priority: number` to resolve ambiguities?

**Comment by Tyriar** [comment]
@rodolf0 the link matcher system used a priority, link providers priority is defined by the order they are registered, so it's in complete control of the embedding application. That's called out here:

https://github.com/xtermjs/xterm.js/blob/1168c6c52f49c7155094fe173ac037ea0e9ce8ac/typings/xterm.d.ts#L754-L755

**Comment by LabhanshAgrawal** [comment]
@rodolf0
maybe you can take a look at https://github.com/LabhanshAgrawal/xterm-link-provider

**Comment by rodolf0** [comment]
> @rodolf0
> maybe you can take a look at [LabhanshAgrawal/xterm-link-provider](https://github.com/LabhanshAgrawal/xterm-link-provider)

Hey @LabhanshAgrawal  this is super cool! Per my understanding the callback is invoked with a range and the text matched correct?

Is there a chance to also invoke the callback with the returned match object from `regexp.exec`? This allows the callback to do fancier replacements. For example, you can use the match groups to extract bits of the text and write an action based on that.

**Comment by LabhanshAgrawal** [comment]
I'm not sure if I understand your example fully
But that callback is provided by xterm, it's not the handler that you provide during creating the class. Your handler will be executed when the link is clicked and the args are provided by xterm i.e. a mouse event and the text from the link.
